formalize component value definitions (#336)
WebAssembly/component-model: design/mvp/Explainer.md

                | <functype>
                | <componenttype>
                | <instancetype>
                | (value <valuebound>) 🪙
                | (type <typebound>)
typebound     ::= (eq <typeidx>)
                | (sub resource)
valuebound    ::= (eq <valueidx>)

>                 | <valtype>

Author: should these be gated by 🪙 as well?

Reviewer: Yep

See the [CanonicalABI.md](CanonicalABI.md#canonical-definitions) for detailed
definitions of each of these built-ins and their interactions.

### 🪙 Value Definitions

Value definitions (in the value index space) are like immutable `global` definitions
in Core WebAssembly except that validation requires them to be consumed exactly
once at instantiation-time (i.e., they are [linear]).

Components may define values in the value index space using following syntax:

```ebnf
value ::= (value <id>? <valtype> <val>)
val   ::= false | true
        | <core:i64>

>         | <core:f64> | NaN

Reviewer: If I'm reading the [text format rules for `f64`](https://webassembly.github.io/spec/core/text/values.html#text-float) correctly, `nan` is already included.  So is `nan:0x____`, for specifying non-canonical NaNs, so we need to disallow that somehow.  Maybe use `<f64canon>` here, defined below as:
```
f64canon ::= <core:f64> without the `nan:0x` case.
```

Author: updated, also updated `Binary.md` to match `nan` casing https://github.com/WebAssembly/component-model/pull/336/commits/302f90a7fc1e670f72b28ea665ae57eb127154ee

Reviewer: Thanks!  As an ultra-nit, could you fix the indentation so that the `::=` of `f64canon` lines up?

        | '<core:char>'
        | <core:name>
        | (record <val>+)
        | (variant "<label>" <val>?)
        | (list <val>*)
        | (tuple <val>+)
        | (flags "<label>"*)
        | (enum "<label>")
        | none | (some <val>)
        | ok | (ok <val>) | error | (error <val>)
```

> 
> Where `val` definition must match the `valtype` specified.
> 
> For example:

Reviewer: ```suggestion
The validation rules for `value` require the `val` to match the `valtype`.  For example:
```

See the [CanonicalABI.md](CanonicalABI.md#canonical-definitions) for detailed
definitions of each of these built-ins and their interactions.

### 🪙 Value Definitions

Value definitions (in the value index space) are like immutable `global` definitions
in Core WebAssembly except that validation requires them to be consumed exactly
once at instantiation-time (i.e., they are [linear]).

Components may define values in the value index space using following syntax:

```ebnf
value ::= (value <id>? <valtype> <val>)
val   ::= false | true
        | <core:u8>
        | <core:s8>
        | <core:s16>
        | <core:u16>
        | <core:s32>
        | <core:u32>
        | <core:s64>
        | <core:u64>
        | <core:f32>
        | <core:f64>
        | '<core:char>'
        | "<core:name>"

>         | (record <val>+)

Author: My initial approach did not include `record`/`tuple` etc. keywords (but did surround in parentheses), however in conversation with @lukewagner I found out that (paraphrasing Luke) by convention `(operator <expr> ... <expr>)` s-expressions are used in place of pure `(<expr> ... <expr>)` in WAT - which is why I've added the type.

It seems to me that omission of parentheses here would introduce ambiguity in parsing e.g. `(record (field "a" (list u8)) (field "b" (list u8)))` values - without some kind of delimiter, how would one differentiate `{a:[1], b:[]}` from `{a:[], b:[1]}` ?

>         | (record <val>+)

Reviewer 2: Ah, I hadn't thought through nesting. I see now that `record`s would get tricky to read, and nested `list`s would create ambiguity. In light of that, In theory we could just omit the parens+keyword for `flags`, `variant`, and `enum`, but then it'd be awkwardly inconsistent with `record`/`list`. So your current proposed syntax here sounds good.

        | (variant "<label>" <val>?)
        | (list <val>*)
        | (tuple <val>+)
        | (flags "<label>"*)
        | (enum "<label>")
        | none | (some <val>)
        | ok | (ok <val>) | error | (error <val>)
```

> Where `val` definition must match the `valtype` specified as defined in [Binary.md](Binary.md).

Reviewer: I'm not sure what you mean here; it seems like Explainer.md shouldn't need to refer to Binary.md (only the other way around).

For example:
```wasm
(component
    (value $a bool true)
    (value $b u8 1)
    (value $c u16 2)
    (value $d u32 3)
    (value $e u64 4)
    (value $f u8 5)
    (value $g u16 6)
    (value $h u32 7)
    (value $i u64 8)
    (value $j f32 9.1)
    (value $k f64 9.2)
    (value $l char 'a')
    (value $m string "hello")
    (value $n (record (field "a" bool) (field "b" u8)) (record true 1))
    (value $o (variant (case "a" bool) (case "b" u8)) (variant "b" 1))
    (value $p (list (result (option u8)))
        (list
            error
            (ok (some 1))
            (ok none)
            error
            (ok (some 2))
        )
    )
    (value $q (tuple u8 u16 u32) (tuple 1 2 3))

    (type $abc (flags "a" "b" "c"))
    (value $r $abc (flags "a" "c"))

    (value $s (enum "a" "b" "c") (enum "b"))

    (type $complex
        (tuple
            (record
                (field "a" (option string))
                (field "b" (tuple (option u8) string))
            )
            (list char)
            $abc
            string
        )
    )

>     (value $complex1 $complex

Reviewer: ```suggestion
    (value $complex1 (type $complex)
```

        (tuple
            (record
                none
                (tuple none "empty")
            )
            (list)
            (flags)
            ""
        )
    )

>     (value $complex2 $complex

Reviewer: ```suggestion
    (value $complex2 (type $complex)
```

See the [CanonicalABI.md](CanonicalABI.md#canonical-definitions) for detailed
definitions of each of these built-ins and their interactions.

### 🪙 Value Definitions

Value definitions (in the value index space) are like immutable `global` definitions
in Core WebAssembly except that validation requires them to be consumed exactly
once at instantiation-time (i.e., they are [linear]).

Components may define values in the value index space using following syntax:

```ebnf
value    ::= (value <id>? <valtype> <val>)
val      ::= false | true
           | <core:i64>
           | <f64canon>
           | '<core:char>'
           | <core:name>
           | (record <val>+)
           | (variant "<label>" <val>?)
           | (list <val>*)
           | (tuple <val>+)
           | (flags "<label>"*)
           | (enum "<label>")
           | none | (some <val>)
           | ok | (ok <val>) | error | (error <val>)

> f64canon ::= <core:f64> without the `nan:0x` case.

Reviewer 3: I think this may want to be sync'd with the `Binary.md` section above? For example integers have discriminants and flags/variants/enums don't have strings for payloads but instead integers.

Author: Will update the numeric types, however the intention for flags, variants and enums is indeed to use labels and not integers

Author: See also https://github.com/WebAssembly/component-model/pull/336#discussion_r1599858948 for potentially an alternative numeric representation with label syntax being the sugar for it

See the [CanonicalABI.md](CanonicalABI.md#canonical-definitions) for detailed
definitions of each of these built-ins and their interactions.

### 🪙 Value Definitions

Value definitions (in the value index space) are like immutable `global` definitions
in Core WebAssembly except that validation requires them to be consumed exactly
once at instantiation-time (i.e., they are [linear]).

Components may define values in the value index space using following syntax:

```ebnf
value    ::= (value <id>? <valtype> <val>)
val      ::= false | true
           | <core:i64>
           | <f64canon>
           | nan
           | '<core:stringchar>'
           | <core:name>
           | (record <val>+)
           | (variant "<label>" <val>?)
           | (list <val>*)
           | (tuple <val>+)
           | (flags "<label>"*)
           | (enum "<label>")
           | none | (some <val>)
           | ok | (ok <val>) | error | (error <val>)
           | (binary <core:datastring>)
f64canon ::= <core:f64> without the `nan:0x` case.
```

The validation rules for `value` require the `val` to match the `valtype`.

> `(binary ...)` expression provides an alternative syntax allowing the binary contents

Reviewer: ```suggestion
The `(binary ...)` expression form provides an alternative syntax allowing the binary contents
```

See the [CanonicalABI.md](CanonicalABI.md#canonical-definitions) for detailed
definitions of each of these built-ins and their interactions.

### 🪙 Value Definitions

Value definitions (in the value index space) are like immutable `global` definitions
in Core WebAssembly except that validation requires them to be consumed exactly
once at instantiation-time (i.e., they are [linear]).

Components may define values in the value index space using following syntax:

```ebnf
value    ::= (value <id>? <valtype> <val>)
val      ::= false | true

>            | <core:u8>

Reviewer: Since we're in the text format here and back to assuming we have a target type, I think we only need `<core:u64>` and `<core:s64>`, and values too big for the target integer type will be rejected as a normal part of validation.

>            | <core:u8>
>            | <core:u16>
>            | <core:u32>

Reviewer: ```suggestion
```

           | <core:u64>

>            | <core:s8>
>            | <core:s16>
>            | <core:s32>

Reviewer: ```suggestion
```

           | <core:s64>

>            | <f32canon>
>            | nan

Reviewer: ```suggestion
```

           | <f64canon>
           | nan
           | '<core:stringchar>'
           | <core:name>
           | (record <val>+)
           | (variant "<label>" <val>?)
           | (list <val>*)
           | (tuple <val>+)
           | (flags "<label>"*)
           | (enum "<label>")
           | none | (some <val>)
           | ok | (ok <val>) | error | (error <val>)
           | (binary <core:datastring>)

> f32canon ::= <core:f32> without the `nan:0x` case.

Reviewer: ```suggestion
```

f64canon ::= <core:f64> without the `nan:0x` case.
```

> The validation rules for `value` require the `val` to match the `valtype`.  For example:

Reviewer: Could you add a sentence that the `(binary ...)` expression provides an alternative syntax allowing the binary contents of the value definition to be written directly in the text format, analogous to data segments, avoiding the need to understand type information when encoding or decoding?

As this example suggests, value imports can serve as generalized [environment
variables], allowing not just `string`, but the full range of `valtype`.

Values can also be exported:
```wasm

> (value $default_url string "https://example.com")
> (export "default-url" (value $default_url))

Reviewer: ```suggestion
(component
  (value $default_url string "https://example.com")
  (export "default-url" (value $default_url))
)
```

```

`default-url` value export will infer `(eq $default_url)` value bound in the component type.

The `eq` bound says that the imported/exported value must be structurally equal
to some preceding value definition. This allows, for example, an imported abstract value to be re-exported.

Include a type ascription to explicitly set type to `string`:

```
(value $default_url string "https://example.com")
(export "default-url" (value $default_url) (value string))

> ```

Reviewer: ````suggestion
Values can also be exported.  For example:
```wasm
(component
  (import "system-port" (value $port u16))
  (value $url string "https://example.com")
  (export "default-url" (value $url))
  (export "default-port" (value $port))
)
```
The inferred type of this component is:
```wasm
(component
  (import "system-port" (value $port u16))
  (value $url string "https://example.com")
  (export "default-url" (value (eq $url)))
  (export "default-port" (value (eq $port)))
)
```
Thus, by default, the precise constant or import being exported is propagated into the component's type and thus its public interface.  In this way, value exports can act as semantic configuration data provided by the component to the host or other client tooling.

Components can also keep the exact value being exported abstract (so that the precise value is not part of the type and public interface) using the "type ascription" feature mentioned [below](#imports-and-exports).
````

Reviewer: And then could we add an example below (where the type ascription feature is introduced) of a value type ascription?

Author: Move the example to the import and export section and other than fixing the section reference and making it consistent with previous occurence applied the diff verbatim

As this example suggests, value imports can serve as generalized [environment
variables], allowing not just `string`, but the full range of `valtype`.

> Example of a value export:

Reviewer: ```suggestion
Values can also be exported:
```

```wasm
(value $default_url string "https://example.com")
(export "default-url" (value $default_url))
;; `default-url` value export will infer `(eq $default_url)` value bound in the component type,
;; include a type ascription to explicitly set type to `string`:
(export "default-url-string" (value $default_url) (value string))

> ```

Reviewer: Could you end the triple-quote block and put the words in comments into a new paragraph, and then start a new triple-quote block with the alternative `export`?

types. In particular, it ensures that a client of a component is able to
externally define a type compatible with the exports of the component.

### 🪙 Value Definitions

Components may define values in the value index
```ebnf
value ::= t:<valtype> v:<val(t)> => (value t v)

> ```

Reviewer: I think this part is in-progress, but just fwiw, I think we also want to be able to optionally bind an identifier, so you'll want `(value <id>? <valtype> <val>)`.  Note also that while Binary.md uses "attribute grammar" syntax (mapping bytes to an AST), here we just want a grammar (no mapping).

So I had been thinking we'd connect this to default parameters, but I guess an even more minimalist first usage of Value Definitions is just defining constant values that are exported, which could actually be useful.  So perhaps show an example component that defines and exports a value.

Oh, and one other really-cool thing is if you could an an `eq` bound to value imports and exports (similar to what we already have in `type` imports and exports) where the right-hand side of the `eq` bound could be a `<validx>`, and that way constant values would become part of the component *type*, which I was actually seeing a use case for in CNCF land (it could be how a component declares what port it wants to bind to without needing a separate config file).

[Index Space]: https://webassembly.github.io/spec/core/syntax/modules.html#indices
[Abbreviations]: https://webassembly.github.io/spec/core/text/conventions.html#abbreviations

[`core:s8`]: https://webassembly.github.io/spec/core/syntax/values.html#integers
[`core:u8`]: https://webassembly.github.io/spec/core/syntax/values.html#integers
[`core:s16`]: https://webassembly.github.io/spec/core/syntax/values.html#integers
[`core:u16`]: https://webassembly.github.io/spec/core/syntax/values.html#integers
[`core:s32`]: https://webassembly.github.io/spec/core/syntax/values.html#integers
[`core:u32`]: https://webassembly.github.io/spec/core/syntax/values.html#integers
[`core:s64`]: https://webassembly.github.io/spec/core/syntax/values.html#integers
[`core:u64`]: https://webassembly.github.io/spec/core/syntax/values.html#integers

> [`core:f32`]: https://webassembly.github.io/spec/core/syntax/values.html#floating-point

Reviewer: Based on the above comment, I think all of these go away and you only keep `f64` and `i64` (note: `i64` contains both `u64` and `s64`).
